subxt/src/backend/legacy/mod.rs

```diff
@@ -77,190 +79,299 @@ impl<T: Config> LegacyBackend<T> {
     }
 
     async fn block_body(&self, at: T::Hash) -> Result<Option<Vec<Vec<u8>>>, Error> {
-        let Some(details) = self.methods.chain_get_block(Some(at)).await? else {
-            return Ok(None);
-        };
-        Ok(Some(
-            details.block.extrinsics.into_iter().map(|b| b.0).collect(),
-        ))
+        retry(|| async {
+            let Some(details) = self.methods.chain_get_block(Some(at)).await? else {
+                return Ok(None);
+            };
+            Ok(Some(
+                details.block.extrinsics.into_iter().map(|b| b.0).collect(),
+            ))
+        })
+        .await
     }
 
     async fn latest_finalized_block_ref(&self) -> Result<BlockRef<T::Hash>, Error> {
-        let hash = self.methods.chain_get_finalized_head().await?;
-        Ok(BlockRef::from_hash(hash))
+        retry(|| async {
+            let hash = self.methods.chain_get_finalized_head().await?;
+            Ok(BlockRef::from_hash(hash))
+        })
+        .await
     }
 
     async fn current_runtime_version(&self) -> Result<RuntimeVersion, Error> {
-        let details = self.methods.state_get_runtime_version(None).await?;
-        Ok(RuntimeVersion {
-            spec_version: details.spec_version,
-            transaction_version: details.transaction_version,
+        retry(|| async {
+            let details = self.methods.state_get_runtime_version(None).await?;
+            Ok(RuntimeVersion {
+                spec_version: details.spec_version,
+                transaction_version: details.transaction_version,
+            })
         })
+        .await
     }
 
     async fn stream_runtime_version(&self) -> Result<StreamOfResults<RuntimeVersion>, Error> {
-        let sub = self.methods.state_subscribe_runtime_version().await?;
-        let sub = sub.map(|r| {
-            r.map(|v| RuntimeVersion {
-                spec_version: v.spec_version,
-                transaction_version: v.transaction_version,
-            })
-        });
-        Ok(StreamOf(Box::pin(sub)))
+        async fn subscribe_runtime_version<T: Config>(
+            methods: &LegacyRpcMethods<T>,
+        ) -> Result<StreamOfResults<RuntimeVersion>, Error> {
+            let sub = methods.state_subscribe_runtime_version().await?;
+            let sub = sub.map(|r| {
+                r.map(|v| RuntimeVersion {
+                    spec_version: v.spec_version,
+                    transaction_version: v.transaction_version,
+                })
+            });
+            Ok(StreamOf(Box::pin(sub)))
+        }
+
+        let sub = retry(|| subscribe_runtime_version(&self.methods)).await?;
+        let methods = self.methods.clone();
+
+        let retry_sub = RetrySubscription::new(
+            sub,
+            Box::new(move || {
+                let methods = methods.clone();
+                Box::pin(async move { subscribe_runtime_version(&methods).await })
+            }),
+        );
```

Comment:
I'm currently a bit confused by the `RetrySubscription` API, in that we:

1. create a stream wrapped in `retry`, and then
2. pass that to the `RetrySubscription`, along with _another_ future that will be ran to restart the thing that's already going to be restarted via `retry`?

I wonder whether you could instead have two methods:

- `retry(|| fut)`; the one you have already which calls the closure to retry a fut whenever disconnected, and
- `retry_stream(|| stream)`; return items from the stream, and call the closure to create a new stream whenever disconnected (could accept a closure that returns a future to create the stream to be more useful).

So then the utils exposed are nice and simple (just two functions) and the code could instead be more like:

```rust
let retry_sub = retry_stream({
    let methods = methods.clone();
    move || subscribe_runtime_version(&methods).await
});
```

Maybe a crappy idea but I'm just wanting to remove the duplication where we are creating the same stream in two places each time :)

Reply:
I get your point it's horrible to read :)